packages/steps/src/StepsConfigParser.ts

```diff
@@ -0,0 +1,218 @@
+import assert from 'node:assert';
+
+import {
+  FunctionStep,
+  isStepFunctionStep,
+  isStepShellStep,
+  ShellStep,
+  Step,
+  validateSteps,
+} from '@expo/eas-build-job';
+
+import { BuildFunction, BuildFunctionById } from './BuildFunction.js';
+import {
+  BuildFunctionGroup,
+  BuildFunctionGroupById,
+  createBuildFunctionGroupByIdMapping,
+} from './BuildFunctionGroup.js';
+import { BuildStepGlobalContext } from './BuildStepContext.js';
+import { BuildStep } from './BuildStep.js';
+import { AbstractConfigParser } from './AbstractConfigParser.js';
+import { BuildConfigError } from './errors.js';
+import { BuildStepOutput } from './BuildStepOutput.js';
+
+export class StepsConfigParser extends AbstractConfigParser {
+  private readonly steps: Step[];
+
+  constructor(
+    ctx: BuildStepGlobalContext,
+    {
+      steps,
+      externalFunctions,
+      externalFunctionGroups,
+    }: {
+      steps: Step[];
+      externalFunctions?: BuildFunction[];
+      externalFunctionGroups?: BuildFunctionGroup[];
+    }
+  ) {
+    super(ctx, {
+      externalFunctions,
+      externalFunctionGroups,
+    });
+
+    this.steps = steps;
+  }
+
+  protected async parseConfigToBuildStepsAndBuildFunctionByIdMappingAsync(): Promise<{
+    buildSteps: BuildStep[];
+    buildFunctionById: BuildFunctionById;
+  }> {
+    const validatedSteps = validateSteps(this.steps);
+    StepsConfigParser.validateAllFunctionsExist(validatedSteps, {
+      externalFunctionIds: this.getExternalFunctionFullIds(),
+      externalFunctionGroupIds: this.getExternalFunctionGroupFullIds(),
```

Comment:
I expect changing custom jobs in a way that does not require validation beforehand, but validates as it runs would be much simpler. Then all these parsers would magically go away?